File: Count Substrings with Exactly K Distinct Characters.c
#include <stdio.h>
#include <string.h>

int atMostK(char *s, int k) {
    int freq[26] = {0};
    int left = 0, count = 0, distinct = 0;

    for (int right = 0; s[right]; right++) {
        if (freq[s[right] - 'a'] == 0)
            distinct++;
        
        freq[s[right] - 'a']++;

        while (distinct > k) {
            freq[s[left] - 'a']--;
            if (freq[s[left] - 'a'] == 0)
                distinct--;
            left++;
        }

        count += right - left + 1;
    }

    return count;
}

int countSubstrings(char *s, int k) {
    return atMostK(s, k) - atMostK(s, k - 1);
}

int main() {
    char s[1000];
    int k;
    scanf("%s %d", s, &k);

    printf("%d", countSubstrings(s, k));
    return 0;
}

File: Count Substrings with Exactly K Distinct Characters.py
def at_most_k(s, k):
    freq = {}
    left = 0
    count = 0

    for right in range(len(s)):
        freq[s[right]] = freq.get(s[right], 0) + 1

        while len(freq) > k:
            freq[s[left]] -= 1
            if freq[s[left]] == 0:
                del freq[s[left]]
            left += 1

        count += right - left + 1

    return count

def count_substrings(s, k):
    return at_most_k(s, k) - at_most_k(s, k - 1)

s = input()
k = int(input())

print(count_substrings(s, k))
